src/osp/Active/SysRender.cpp

```diff
@@ -77,57 +70,78 @@ void SysRender::setup(ActiveScene& rScene)
         depthStencil->setStorage(GL::RenderbufferFormat::Depth24Stencil8, viewSize);
 
         DependRes<GL::Framebuffer> fbo =
-            resources.add<GL::Framebuffer>("offscreen_fbo", Range2Di{{0, 0}, viewSize});
+            rCtxResources.add<GL::Framebuffer>("offscreen_fbo", Range2Di{{0, 0}, viewSize});
         fbo->attachTexture(GL::Framebuffer::ColorAttachment{0}, *color, 0);
         fbo->attachRenderbuffer(GL::Framebuffer::BufferAttachment::DepthStencil, *depthStencil);
+    }
+}
 
 
-        rScene.reg_emplace<ACompRenderTarget>(rScene.hier_get_root(),
-            viewSize, std::move(fbo));
-        rScene.reg_emplace<ACompFBOColorAttachment>(rScene.hier_get_root(), color);
-    }
+void SysRender::setup_forward_renderer(ActiveScene &rScene)
+{
+    Package &rCtxResources = rScene.get_context_resources();
+
+    Vector2i viewSize = Magnum::GL::defaultFramebuffer.viewport().size();
 
-    // Create the default render pipeline
-    resources.add<RenderPipeline>("default", create_forward_renderer());
+    rScene.reg_emplace<ACompRenderTarget>(
+            rScene.hier_get_root(), viewSize,
+            rCtxResources.get<Magnum::GL::Framebuffer>("offscreen_fbo"));
+    rScene.reg_emplace<ACompFBOColorAttachment>(
+            rScene.hier_get_root(),
+            rCtxResources.get<Magnum::GL::Texture2D>("offscreen_fbo_color"));
+
+    // Add render groups
+    auto &rGroups = rScene.get_registry().set<ACtxRenderGroups>();
+    rGroups.m_groups.emplace("fwd_opaque", RenderGroup{});
+    rGroups.m_groups.emplace("fwd_transparent", RenderGroup{}).second;
+
+    // Create the default render pipeline (if not already added yet)
+    rCtxResources.add<RenderPipeline>("default", create_forward_pipeline(rScene));
 }
 
-RenderPipeline SysRender::create_forward_renderer()
+RenderPipeline SysRender::create_forward_pipeline(ActiveScene& rScene)
 {
     /* Define render passes */
-    RenderSteps_t pipeline;
+    std::vector<RenderStep_t> pipeline;
 
     // Opaque pass
     pipeline.emplace_back(
         [](ActiveScene& rScene, ACompCamera const& camera)
         {
-            using namespace Magnum;
             using Magnum::GL::Renderer;
 
-            auto& reg = rScene.get_registry();
-
             Renderer::enable(Renderer::Feature::DepthTest);
             Renderer::enable(Renderer::Feature::FaceCulling);
             Renderer::disable(Renderer::Feature::Blending);
-
-            // Fetch opaque objects
-            auto opaqueView = reg.view<ACompOpaque, ACompVisible, ACompShader>();
-
-            SysRender::draw_group(rScene, opaqueView, camera);
+            Renderer::setDepthMask(true);
+
+            ActiveReg_t &rReg = rScene.get_registry();
+            auto const viewVisible = rReg.view<const ACompVisible>();
+            auto &rGroups = rReg.ctx<ACtxRenderGroups>();
+
+            for (auto const& [ent, toDraw]
+                 : rGroups.m_groups["fwd_opaque"].view().each())
```

Comment:
is your intention to create-if-not-exists ?

Imho, that should be happening before you get to the render pipeline.

Since the render pipeline is going to execute the same instructions over and over, no sense in paying the cost of checking if the thing is already allocated and allocating if not, over and over.

Reply:
The intention is to always get, where the key always exists. This will be changed to at()